File: browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.cc
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "mises/browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.h"

#include <utility>

#include "mises/browser/brave_wallet/brave_wallet_tab_helper.h"
#include "chrome/browser/ui/browser_finder.h"
#include "content/public/browser/web_contents.h"

namespace {

base::OnceCallback<void()> g_new_setup_needed_callback_for_testing;
base::OnceCallback<void()> g_account_creation_callback_for_testing;

}  // namespace

namespace brave_wallet {

void ShowPanel(content::WebContents* web_contents) {
  if (!web_contents)
    return;

  auto* tab_helper =
      brave_wallet::BraveWalletTabHelper::FromWebContents(web_contents);
  if (tab_helper)
    tab_helper->ShowBubble();
}
void ShowApprovePanel(content::WebContents* web_contents) {
  if (!web_contents)
    return;

  auto* tab_helper =
      brave_wallet::BraveWalletTabHelper::FromWebContents(web_contents);
  if (tab_helper)
    tab_helper->ShowApproveWalletBubble();
}
void ClosePanel(content::WebContents* web_contents) {
  
  if (!web_contents)
    return;

  auto* tab_helper =
      brave_wallet::BraveWalletTabHelper::FromWebContents(web_contents);
  if (tab_helper)
    tab_helper->CloseBubble();
}
bool IsPanelShowing(content::WebContents* web_contents) {
  if (!web_contents)
    return false;

  auto* tab_helper =
      brave_wallet::BraveWalletTabHelper::FromWebContents(web_contents);
  if (!tab_helper)
    return false;
  return tab_helper->IsShowingBubble();
}

void ShowWalletOnboarding(content::WebContents* web_contents) {
  Browser* browser =
      web_contents ? chrome::FindBrowserWithTab(web_contents) : nullptr;

  if (browser) {
    //brave::ShowBraveWalletOnboarding(browser);
  } else if (g_new_setup_needed_callback_for_testing) {
    std::move(g_new_setup_needed_callback_for_testing).Run();
  }
}

void ShowAccountCreation(content::WebContents* web_contents,
                         const std::string& keyring_id) {
  Browser* browser =
      web_contents ? chrome::FindBrowserWithTab(web_contents) : nullptr;

  if (browser) {
    //brave::ShowBraveWalletAccountCreation(browser, keyring_id);
  } else if (g_account_creation_callback_for_testing) {
    std::move(g_account_creation_callback_for_testing).Run();
  }
}

void WalletInteractionDetected(content::WebContents* web_contents) {}

// Desktop uses a panel to show all notifications instead of a dialog
// on Android for permissions
bool IsWeb3NotificationAllowed() {
  return true;
}

void SetCallbackForNewSetupNeededForTesting(
    base::OnceCallback<void()> callback) {
  g_new_setup_needed_callback_for_testing = std::move(callback);
}

void SetCallbackForAccountCreationForTesting(
    base::OnceCallback<void()> callback) {
  g_account_creation_callback_for_testing = std::move(callback);
}

}  // namespace brave_wallet

File: android/java/org/chromium/chrome/browser/crypto_wallet/BraveWalletProviderDelegateImplHelper.java
/* Copyright (c) 2022 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.crypto_wallet;
import android.app.Activity;

import org.chromium.base.Callback;
import org.chromium.base.Log;
import org.jni_zero.CalledByNative;
import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;
import org.chromium.content_public.browser.WebContents;
import org.chromium.base.Callback;
import org.chromium.base.Callbacks;
import org.chromium.chrome.browser.customtabs.CustomTabActivity;
import org.chromium.base.MisesSysUtils;
import org.chromium.base.ApplicationStatus;
import org.chromium.chrome.browser.tab.Tab;
import org.chromium.chrome.browser.ChromeTabbedActivity;


@JNINamespace("brave_wallet")
public class BraveWalletProviderDelegateImplHelper {
    private static final String TAG = "BraveWalletProvider";

    @CalledByNative
    public static void showPanel(final String url) {
      if (MisesSysUtils.getActivityContext() == null) {
        return;
      }
      
      try {
          BraveWalletCustomTabActivity.showWalletPage(MisesSysUtils.getActivityContext(),url);
      } catch (Exception e) {
          Log.e(TAG, "showPanel " + e);
      }
    }

    @CalledByNative
    public static void closePanel() {
      if (MisesSysUtils.getActivityContext() == null) {
        return;
      }
      
      try {
        for (Activity activity : ApplicationStatus.getRunningActivities()) {
            if (!(activity instanceof BraveWalletCustomTabActivity)) continue;

            BraveWalletCustomTabActivity customTabActivity = (BraveWalletCustomTabActivity) activity;
            customTabActivity.finishAndRemoveTask();

            
        }
      } catch (Exception e) {
          Log.e(TAG, "closePanel " + e);
      }
    }

    @CalledByNative
    public static boolean isPanelShowing() {
      if (MisesSysUtils.getActivityContext() == null) {
        return false;
      }
      boolean showing = false;
      try {
        for (Activity activity : ApplicationStatus.getRunningActivities()) {
            if (!(activity instanceof BraveWalletCustomTabActivity)) continue;
            if (!activity.isFinishing()) {
              showing = true;
              break;
            }
        }
      } catch (Exception e) {
          Log.e(TAG, "isPanelShowing " + e);
      }

      return showing;
    }

    @CalledByNative
    public static void showWalletOnboarding() {
        // try {
        //     BraveActivity activity = BraveActivity.getBraveActivity();
        //     activity.showWalletOnboarding();
        // } catch (BraveActivity.BraveActivityNotFoundException e) {
        //     Log.e(TAG, "showWalletOnboarding " + e);
        // }

        
    }

    @CalledByNative
    public static void walletInteractionDetected(WebContents webContents) {
        // try {
        //     BraveActivity activity = BraveActivity.getBraveActivity();
        //     activity.walletInteractionDetected(webContents);
        // } catch (BraveActivity.BraveActivityNotFoundException e) {
        //     Log.e(TAG, "walletInteractionDetected " + e);
        // }
    }

    @CalledByNative
    public static boolean isWeb3NotificationAllowed() {
        return true;
    }

    @CalledByNative
    public static void ShowAccountCreation(String keyringId) {
        // try {
        //     BraveActivity activity = BraveActivity.getBraveActivity();
        //     activity.showAccountCreation(keyringId);
        // } catch (BraveActivity.BraveActivityNotFoundException e) {
        //     Log.e(TAG, "ShowAccountCreation " + e);
        // }
    }

    public static void IsSolanaConnected(
            WebContents webContents, String account, Callbacks.Callback1<Boolean> callback) {
        Callback<Boolean> callbackWrapper =
                result -> {
                    callback.call(result);
                };
        BraveWalletProviderDelegateImplHelperJni.get()
                .IsSolanaConnected(webContents, account, callbackWrapper);
    }

    public static void OnWalletPanelClosed() {
      Log.i(TAG, "OnWalletPanelClosed");
      Activity context = MisesSysUtils.getActivityContext();
      if (context == null || !(context instanceof ChromeTabbedActivity)) {
        return;
      }
      Tab tab = ((ChromeTabbedActivity)context).getActivityTab();
      WebContents webContents = tab == null ? null : tab.getWebContents();
      if (webContents == null) {
        return;
      }
      if (!isPanelShowing()) {
        BraveWalletProviderDelegateImplHelperJni.get().OnWalletPanelClosed(webContents);
      }
      
    }

    @NativeMethods
    interface Natives {
        void IsSolanaConnected(WebContents webContents, String account, Callback<Boolean> callback);
        void OnWalletPanelClosed(WebContents webContents);
    }
}
